// @mfd/rbxdatastoreservice/Bin/Enumeration/ErrorType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorType = void 0;
/**
 * @internal
 */
var ErrorType;
(function (ErrorType) {
    ErrorType[ErrorType["NO_EMPTY_KEYNAME"] = 101] = "NO_EMPTY_KEYNAME";
    ErrorType[ErrorType["KEYNAME_TOO_LARGE"] = 102] = "KEYNAME_TOO_LARGE";
    ErrorType[ErrorType["DATA_NOT_ALLOWED_IN_DATASTORE"] = 103] = "DATA_NOT_ALLOWED_IN_DATASTORE";
    ErrorType[ErrorType["CANNOT_STORE_DATA_IN_DATASTORE"] = 104] = "CANNOT_STORE_DATA_IN_DATASTORE";
    ErrorType[ErrorType["VALUE_TOO_LARGE"] = 105] = "VALUE_TOO_LARGE";
    ErrorType[ErrorType["MAX_VAL_AND_MIN_VAL_NOT_INTEGERS"] = 106] = "MAX_VAL_AND_MIN_VAL_NOT_INTEGERS";
    ErrorType[ErrorType["PAGE_SIZE_MUST_BE_IN_RANGE"] = 107] = "PAGE_SIZE_MUST_BE_IN_RANGE"; /* =106 */
    ErrorType[ErrorType["NO_API_ACCESS_ALLOWED"] = 108] = "NO_API_ACCESS_ALLOWED";
    ErrorType[ErrorType["ORDERERD_DATATORE_DELETED"] = 109] = "ORDERERD_DATATORE_DELETED";
    ErrorType[ErrorType["CANNOT_PARSE_RESPONSE"] = 110] = "CANNOT_PARSE_RESPONSE";
    ErrorType[ErrorType["API_SERVICES_REJECTED"] = 111] = "API_SERVICES_REJECTED";
    ErrorType[ErrorType["KEY_NOT_FOUND"] = 112] = "KEY_NOT_FOUND";
    ErrorType[ErrorType["MALFORMED_DATASTORE_RESPONSE"] = 113] = "MALFORMED_DATASTORE_RESPONSE";
    ErrorType[ErrorType["MALFORMED_ORDERED_DATASTORE_RESPONSE"] = 114] = "MALFORMED_ORDERED_DATASTORE_RESPONSE";
    ErrorType[ErrorType["METADATA_TOO_LARGE"] = 115] = "METADATA_TOO_LARGE";
    ErrorType[ErrorType["USERID_LIMIT_TOO_LARGE"] = 116] = "USERID_LIMIT_TOO_LARGE";
    ErrorType[ErrorType["USERID_ATTRIBUTE_INVALID"] = 117] = "USERID_ATTRIBUTE_INVALID";
    ErrorType[ErrorType["METATADA_ATTRIBUTE_INVALID"] = 118] = "METATADA_ATTRIBUTE_INVALID";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
