// @mfd/rbxdatastoreservice/Bin/Classes/DataStoreIncrementOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataStoreIncrementOptions = void 0;
/**
 * An object that specifies additional parameters for a [GlobalDataStore:IncrementAsync](https://developer.roblox.com/en-us/api-reference/class/DataStoreIncrementOptions) call.
 *
 * ---
 * SEE ALSO
 * --------
 * - [Data Stores](https://developer.roblox.com/en-us/articles/Data-store), an in-depth guide on data structure, management, error handling, etc.
 *
 * ---
 */
class DataStoreIncrementOptions {
    /**
     * @internal
     */
    constructor(attributes) {
        this.attributes = attributes || {};
    }
    /**
     * This function gets custom metadata associated with this [DataStoreIncrementOptions](https://developer.roblox.com/en-us/api-reference/class/DataStoreIncrementOptions) instance.
     * @returns {Record<string, unknown>} Metadata associated with this [DataStoreIncrementOptions](https://developer.roblox.com/en-us/api-reference/class/DataStoreIncrementOptions) instance.
     */
    GetMetadata() {
        return this.attributes;
    }
    /**
     * This function sets custom metadata used by [GlobalDataStore:IncrementAsync](https://developer.roblox.com/en-us/api-reference/class/DataStoreIncrementOptions) to associate metadata with a key. Metadata should be in key-value pair form.
     *
     * ---
     * SEE ALSO
     * --------
     * - [Data Stores](https://developer.roblox.com/en-us/articles/Data-store), an in-depth guide on data structure, management, error handling, etc.
     *
     * ---
     *
     * @param {Record<string, unknown>} attributes Metadata values to set for the key.
     */
    SetMetadata(attributes) {
        if (this.attributes !== attributes) {
            this.attributes = attributes;
        }
    }
}
exports.DataStoreIncrementOptions = DataStoreIncrementOptions;
